EZP-31211 Document how to add a button for editing content in the Back Office (#947)
ezsystems/developer-documentation: docs/tutorials/extending_admin_ui/3b_adding_edit_button_content_list.md

# Step 3b - Adding Edit button to content list

In this step you will add an edit button to the content list. The following tutorial is based on the [Step 2 - Creating a content list](2_creating_a_content_list.md).

## Modify the controller

Introduce changes to `src/Controller/AllContentListController.php`, so that it allows content editing from the content list.

> First, add the new `FormFactory $formFactory` parameter and the `$this->formFactory = $formFactory;` argument to the `__construct` function:

Reviewer: I think this is too simplified, and to much step by step.
I would go with something along 'Inject EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory service into your controller`, the `use statement` and assigning to variable are kinda self explanatory. 

Unless this is style of whole tutorial, then it is ok.

```php hl_lines="1 5"

> public function __construct(SearchService $searchService, ContentTypeService $contentTypeService, FormFactory $formFactory)

Reviewer: we could format this better
```php
public function __construct(
    SearchService $searchService,
    ContentTypeService $contentTypeService,
    FormFactory $formFactory
) {
    $this->searchService = $searchService;
    $this->contentTypeService = $contentTypeService;
    $this->formFactory = $formFactory;
}
```

{
    $this->searchService = $searchService;
    $this->contentTypeService = $contentTypeService;
    $this->formFactory = $formFactory;
}
```

Next, provide a new use statement for `FormFactory` parameter:

```php
use EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory;
```

> After the lines setting the `$paginator` parameters, add the following code line:

Reviewer: it dosent have to be after those lines, I would also explain what this do - creating underlining form for handling request for content editing.

```php
$editForm = $this->formFactory->contentEdit();
```

> Finally, provide the new parameter to `$this->render`, after `articles`:

Reviewer: does not have to be after `articles`

All below code blocks should be added to `templates/list/all_content_list.html.twig`.

First, add `<th>{{ 'Edit'|trans }}</th>` header to the content list table inside `<section class="container my-4">`:

```html+twig  hl_lines="8"
<table class="table">
    <thead>
    <tr>
        <th>{{ 'Content name'|trans }}</th>
        <th>{{ 'Content Type'|trans }}</th>
        <th>{{ 'Modified'|trans }}</th>
        <th>{{ 'Published'|trans }}</th>
        <th>{{ 'Edit'|trans }}</th>
    </tr>
    </thead>
```

Next, add the edit button as a new <td> tag inside `<section class="container my-4">`:

```html+twig
<td>
    <button class="btn btn-icon mx-2 ez-btn--content-edit"
            title="{{ 'dashboard.table.all.content.edit'|trans|desc('Edit Content') }}"
            data-content-id="{{ article.contentInfo.id }}"
            data-version-no="{{ article.contentInfo.currentVersionNo }}"
            data-language-code="{{ article.contentInfo.mainLanguageCode }}">
        <svg class="ez-icon ez-icon-edit">
            <use xlink:href="{{ asset('bundles/ezplatformadminui/img/ez-icons.svg') }}#edit"></use>
        </svg>
    </button>
</td>
```

> After that, add a hidden form with js listeners for draft conflict checks by adding the following code block inside `{%- block content -%}`, right under `<section class="container my-4">`:

Reviewer: those are two different concepts, form is used for redirecting to content edit page as we dont use `get` request for that kind of users actions.

also this form does not have any js listeners, it is used by them.

listeners are in `{% block javascripts %}` block.

# Step 3b - Adding Edit button to content list

> In this step you will add an edit button to the content list. The following tutorial is based on the [Step 2 - Creating a content list](2_creating_a_content_list.md).

Reviewer 2: ```suggestion
In this step you will add an edit button to the content list. The following tutorial is requires the completed [Step 2 - Creating a content list](2_creating_a_content_list.md).
```

## Modify the controller

> Introduce changes to `src/Controller/AllContentListController.php`, so that it allows content editing from the content list.

Reviewer 2: ```suggestion
Introduce changes to `src/Controller/AllContentListController.php`, so that it enables content editing from the content list.
```

> First, inject `EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory` service into your controller.

Reviewer 2: ```suggestion
First, inject the `EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory` service into your controller.
```

```php
use EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory;
```

Create an underlying form for handling requests for content editing.
Add the following code line after e.g. the lines setting the `$paginator` parameters:

```php
$editForm = $this->formFactory->contentEdit();
```

Finally, provide the new parameter to `$this->render`. It can be added after e.g. `articles`:

``` php
'form_edit' => $editForm->createView(),
```

??? tip "Complete controller code"

    ```php hl_lines="6 20 24 36 40"
    <?php
    
    namespace App\Controller;
    
    use EzSystems\EzPlatformAdminUiBundle\Controller\Controller;
    use EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory;
    use eZ\Publish\API\Repository\SearchService;
    use eZ\Publish\API\Repository\ContentTypeService;
    use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
    use eZ\Publish\API\Repository\Values\Content\LocationQuery;
    use eZ\Publish\Core\Pagination\Pagerfanta\LocationSearchAdapter;
    use Pagerfanta\Pagerfanta;
    
    class AllContentListController extends Controller
    {
        private $searchService;
    
        private $contentTypeService;

>     

Reviewer: please include 
`private $formFactory`

>         public function __construct(SearchService $searchService, ContentTypeService $contentTypeService, FormFactory $formFactory)

Reviewer: whole formating of this snippet is off, mostly in `listAction`

        {
            $this->searchService = $searchService;
            $this->contentTypeService = $contentTypeService;
            $this->formFactory = $formFactory;
        }
    
        public function listAction($page = 1)
        {
            $query = new LocationQuery();        $criterions = [
            new Criterion\Visibility(Criterion\Visibility::VISIBLE),
        ];        $query->query = new Criterion\LogicalAnd($criterions);        $paginator = new Pagerfanta(
            new LocationSearchAdapter($query, $this->searchService)
        );
            $paginator->setMaxPerPage(8);
            $paginator->setCurrentPage($page);
            $editForm = $this->formFactory->contentEdit();
            return $this->render('list/all_content_list.html.twig', [
                'totalCount' => $paginator->getNbResults(),
                'articles' => $paginator,
                'form_edit' => $editForm->createView(),
            ]);
        }
    }
    ```

## Change the template

The last thing to do is to add the edit button to the content list template.

> All below code blocks should be added to `templates/list/all_content_list.html.twig`.

Reviewer 2: ```suggestion
All the code blocks below should be added to `templates/list/all_content_list.html.twig`.
```

> First, add `<th>{{ 'Edit'|trans }}</th>` header to the content list table inside `<section class="container my-4">`:

Reviewer 2: ```suggestion
First, add a `<th>{{ 'Edit'|trans }}</th>` header to the content list table inside `<section class="container my-4">`:
```

```html+twig  hl_lines="8"
<table class="table">
    <thead>
    <tr>
        <th>{{ 'Content name'|trans }}</th>
        <th>{{ 'Content Type'|trans }}</th>
        <th>{{ 'Modified'|trans }}</th>
        <th>{{ 'Published'|trans }}</th>
        <th>{{ 'Edit'|trans }}</th>
    </tr>
    </thead>
```

> Next, add the edit button as a new <td> tag inside `<section class="container my-4">`:

Reviewer 2: ```suggestion
Next, add the edit button as a new `<td>` tag inside `<section class="container my-4">`:
```

```html+twig
<td>
    <button class="btn btn-icon mx-2 ez-btn--content-edit"
            title="{{ 'dashboard.table.all.content.edit'|trans|desc('Edit Content') }}"
            data-content-id="{{ article.contentInfo.id }}"
            data-version-no="{{ article.contentInfo.currentVersionNo }}"
            data-language-code="{{ article.contentInfo.mainLanguageCode }}">
        <svg class="ez-icon ez-icon-edit">
            <use xlink:href="{{ asset('bundles/ezplatformadminui/img/ez-icons.svg') }}#edit"></use>
        </svg>
    </button>
</td>
```

> After that, add a hidden form for redirecting to content edit page by adding the following code block inside `{%- block content -%}`, right under `<section class="container my-4">`:

Reviewer 2: ```suggestion
After that, add a hidden form for redirecting to the content edit page by adding the following code block inside `{%- block content -%}`, right under `<section class="container my-4">`:
```

                                <svg class="ez-icon ez-icon-edit">
                                    <use xlink:href="{{ asset('bundles/ezplatformadminui/img/ez-icons.svg') }}#edit"></use>
                                </svg>
                            </button>
                        </td>
                    </tr>
                {% endfor %}
                </tbody>
            </table>
            {{ pagerfanta(articles, 'ez') }}
        </section>    
        {{ form_start(form_edit, {
            'action': path('ezplatform.content.edit'),
            'attr':
            { 'class': 'ez-edit-content-form'}
        }) }}
        {{ form_widget(form_edit.language, {'attr': {'hidden': 'hidden', 'class': 'language-input'}}) }}
        {{ form_end(form_edit) }}
        {% include '@ezdesign/content/modal/version_conflict.html.twig' %}
    {%- endblock -%}
    
    {% block javascripts %}
        {{ encore_entry_script_tags('ezplatform-admin-ui-dashboard-js', null, 'ezplatform') }}
    {%- endblock -%}
    ```

## Check results

!!! tip

    If you cannot see the results or encounter an error, clear the cache and reload the application.

> At this point you should see the edit button beside each content item in the content list.

Reviewer 2: ```suggestion
At this point you should see the edit button beside each Content item in the content list.
```

> Select the edit button to change your content items.

Reviewer 2: ```suggestion
Select the edit button to change your Content items.
```

Introduce changes to `src/Controller/AllContentListController.php`, so that it allows content editing from the content list.

First, inject `EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory` service into your controller.

To do so, add the new `FormFactory $formFactory` parameter and the `$this->formFactory = $formFactory;` argument to the `__construct` function:

> ```php hl_lines="1 5"

Reviewer 2: ```suggestion
```php hl_lines="4 8"
```

use EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory;
```

> Create underlining form for handling requests for content editing.

Reviewer 2: ```suggestion
Create an underlying form for handling requests for content editing.
```

Add the following code line after e.g. the lines setting the `$paginator` parameters:

```php
$editForm = $this->formFactory->contentEdit();
```

> Finally, provide the new parameter to `$this->render`, it can be added after e.g. `articles`:

Reviewer 2: ```suggestion
Finally, provide the new parameter to `$this->render`. It can be added after e.g. `articles`:
```

{% include '@ezdesign/content/modal/version_conflict.html.twig' %}
```

> Finally, add a javascript block with js listeners at the end of the twig file:

Reviewer 2: ```suggestion
Finally, add a JavaScript block with js listeners at the end of the twig file:
```
